code/game/machinery/ecto_sniffer.dm

```diff
@@ -0,0 +1,83 @@
+/obj/machinery/ecto_sniffer
+	name = "ectoscopic sniffer"
+	desc = "A highly sensitive parascientific instrument calibrated to detect the slightest whiff of ectoplasm."
+	icon = 'icons/obj/machines/research.dmi'
+	icon_state = "ecto_sniffer"
+	density = FALSE
+	anchored = FALSE
+	pass_flags = PASSTABLE
+	circuit = /obj/item/circuitboard/machine/ecto_sniffer
+	///determines if the device if the power switch is turned on or off. Useful if the ghosts are too annoying.
+	var/on = TRUE
+	///If this var set to false the ghosts will not be able interact with the machine, say if the machine is silently disabled by cutting the internal wire.
+	var/sensor_enabled = TRUE
+	///List of ckeys containing players who have recently activated the device, players on this list are prohibited from activating the device untill their residue decays.
+	var/list/ectoplasmic_residues = list()
+
+/obj/machinery/ecto_sniffer/Initialize()
+	. = ..()
+	wires = new/datum/wires/ecto_sniffer(src)
+
+/obj/machinery/ecto_sniffer/attack_ghost(mob/user)
+	. = ..()
+	if(!is_operational() || !on || !sensor_enabled)
+		return
+
+	for(var/spirit_key in ectoplasmic_residues)
+		if(spirit_key == user.ckey)
+			return
```

Comment:
Could this be replaced with user.ckey in ectoplasmic_residues or an assoc list instead to do a dictionary access?
```
ectoplasmic_residues[ckey] = TRUE
if(ectoplasmic_residues[ckey])
   ...
```